<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
</head>
<body>
<h1 id="uniqtag-assign-unique-reasonably-stable-content-derived-identifiers-to-genes">UniqTag: Assign unique, reasonably stable, content-derived identifiers to genes</h1>
<p>Shaun Jackman, Joerg Bohlmann, İnanç Birol</p>
<h1 id="abstract">Abstract</h1>
<h2 id="summary">Summary</h2>
<p>UniqTag assigns unique identifiers to gene sequences, or other arbitrary sequences of characters, that are derived from the <em>k</em>-mer composition of the sequence. Unlike serial or accession numbers, these identifiers are reasonably stable between different assemblies and annotations of the same data.</p>
<h2 id="availability-and-implementation">Availability and implementation</h2>
<p>The implementation of UniqTag is available at <code>https://github.com/sjackman/uniqtag</code></p>
<p>Supplementary data and the code used to generate it is available at <code>https://github.com/sjackman/uniqtag-paper</code></p>
<h2 id="contact">Contact</h2>
<p>Shaun Jackman &lt;sjackman@bcgsc.ca&gt;</p>
<h1 id="introduction">Introduction</h1>
<p>The task of annotating the genes of a genome sequence often follows genome sequence assembly. These annotated genes are assigned unique identifiers by which they can be referenced. Assembly and annotation is often an iterative process, by refining the method or by the addition of the more sequencing data. These gene identifiers would ideally be reasonably stable from one assembly to the next. Genes are typically assigned serial or accession numbers, which, although certainly unique, are not stable between assemblies. A single change in the assembly can result in a total renumbering of the annotated genes.</p>
<p>One solution to stabilize identifiers is to assign them based on the content of the gene sequence. A cryptographic hash function such as SHA (Secure Hash Algorithm) (<a href="http://www.nist.gov/manuscript-publication-search.cfm?pub_id=910977">Dang, 2012</a>) derives a message digest from the sequence, such that two sequences with the same content will have the same message digest, and two sequences that differ will have different message digests. If a cryptographic hash were used to identify a gene, the same gene in two assemblies with identical content would be assigned identical identifiers, but by design a slight change in the sequence, such as a single-character substitution, would result in a completely different digest and identifier.</p>
<p>A cryptographic hash function is designed so that small changes in the message, even a single bit change, results in large changes to the message digest: half of the bits of the digest are expected to flip, called the avalanche effect (<a href="http://www.scientificamerican.com/article/cryptography-and-computer-privacy/">Feistel, 1973</a>). Locality-sensitive hashing (LSH) in contrast aims to assign items that are similar to the same hash bucket. A hash function that, after a small perturbation of the sequence, assigns an identical identifier to the sequence is desirable for identifying the genes of a genome sequence assembly project. One such locality-sensitive hash function, MinHash, was employed in identifying web pages with similar content (<a href="http://dx.doi.org/10.1109/SEQUEN.1997.666900">Broder, 1997</a>). UniqTag implements MinHash, where the set of elements of an item is the <em>k</em>-mer composition of the sequence, the hash function is the identity function and the minimal element is the lexicographically minimal sequence, to assign reasonably stable identifiers to genes. These identifiers are intended for systematic identification, unique within an assembly, rather than as a biological name, which is typically assigned based on biological function or homology to orthologous genes.</p>
<h1 id="algorithm">Algorithm</h1>
<p>The following symbols and terms are defined.</p>
<ul>
<li><span class="math">\(\Sigma\)</span> is an alphabet</li>
<li><em>s</em> and <em>t</em> are strings, a sequence of symbols over <span class="math">\(\Sigma\)</span></li>
<li><em>S</em> is a set of strings over <span class="math">\(\Sigma\)</span></li>
<li><span class="math">\(\Sigma^k\)</span> is the set of all strings over <span class="math">\(\Sigma\)</span> of length <em>k</em></li>
<li><span class="math">\(\min S\)</span> is the lexicographically minimal string of <em>S</em></li>
<li><span class="math">\(C(s)\)</span> is the set of all substrings of <em>s</em></li>
<li>A <em>k</em>-mer of <em>s</em> is a substring of <em>s</em> with length <em>k</em>, also called an <em>n</em>-gram</li>
<li><span class="math">\(C_k(s)\)</span> is the set of all <em>k</em>-mers present in <em>s</em></li>
<li><span class="math">\(f(s, S)\)</span> is the frequency of <em>s</em> in <em>S</em>, defined as the number of strings in <em>S</em> that contain <em>s</em> as a substring</li>
<li><span class="math">\(\mathop{\arg\,\min}\limits_{t \in C_k(s)} f(t, S)\)</span> is the set of the <em>k</em>-mers of <em>s</em> that are least frequent in <em>S</em></li>
<li><span class="math">\(u_k(s, S)\)</span> is the UniqTag, the lexicographically minimal <em>k</em>-mer of the <em>k</em>-mers of <em>s</em> that are least frequent in <em>S</em></li>
</ul>
<p>The UniqTag <span class="math">\(u_k(s, S)\)</span> is defined as follows.</p>
<p><span class="math">\[
\begin{aligned}
C_k(s) &amp;= C(s) \cap \Sigma^k
\\ f(s, S) &amp;= \left\vert \{ s \in C(t) \wedge t \mid t \in S \} \right\vert
\\ u_k(s, S) &amp;= \min \mathop{\arg\,\min}\limits_{t \in C_k(s)} f(t, S)
\end{aligned}
\]</span></p>
<h1 id="discussion">Discussion</h1>
<p>When iterating over multiple assemblies of the same data, it is rather inconvenient when gene identifiers to change from one assembly to the next. UniqTag attempts to address this common annoyance. By identifying the gene by a feature of its content rather than an arbitrary serial number, the gene identifier is reasonably stable between assemblies.</p>
<p>A UniqTag will change due to a difference in the locus of the UniqTag itself, the creation of a least-frequent <em>k</em>-mer that is lexicographically smaller than the previous UniqTag, or the creation of a <em>k</em>-mer elsewhere resulting in the UniqTag no longer being a least-frequent <em>k</em>-mer. Concatenating two gene models results in a gene whose UniqTag is the minimum of the two previous UniqTags, unless one of the k-mer at the junction of the two sequences is lexicographically smaller. Similarly, splitting a gene model in two results in one gene that retains the previous UniqTag and a second gene that is assigned a new UniqTag, unless the previous UniqTag spanned the junction.</p>
<p>A UniqTag can be generated from the nucleotide sequence of a gene. Using instead the translated amino acid sequence of a protein-coding gene sequence results in a UniqTag that is stable across synonymous changes to the coding sequence as well as to changes in the untranslated regions (UTR) and introns of the gene. Since the amino acid alphabet is larger than the nucleotide alphabet, fewer characters are required for a <em>k</em>-mer to be likely unique, resulting in an aesthetically pleasing shorter identifier.</p>
<p>Two gene models with identical sequence will have the same UniqTag. It is possible although uncommon that two non-identical genes with similar <em>k</em>-mer composition and no unique <em>k</em>-mer are assigned the same UniqTag. Genes that are assigned the same UniqTag are distinguished by adding a numerical suffix to the UniqTag.</p>
<h1 id="acknowledgements">Acknowledgements</h1>
<p>Thanks to Nathaniel Street for his enthusiastic feedback, to the SMarTForests project, its funders and the organizers of the 2014 Conifer Genome Summit that made our conversation possible.</p>
<h1 id="funding">Funding</h1>
<p>This work was supported by the Natural Sciences and Engineering Research Council of Canada, Genome British Columbia, Genome Alberta, Genome Québec and Genome Canada.</p>
<h1 id="references">References</h1>
<ul>
<li><a href="http://dx.doi.org/10.1109/SEQUEN.1997.666900">Broder, A. Z. (1997)</a> On the resemblance and containment of documents. <em>Compression and Complexity of Sequences</em>, 1997 Proceedings, 21-29.</li>
<li><a href="http://www.nist.gov/manuscript-publication-search.cfm?pub_id=910977">Dang, Q. H. (2012)</a> Secure Hash Standard (SHS). <em>NIST FIPS</em>, 180(4), 1-35.</li>
<li><a href="http://www.scientificamerican.com/article/cryptography-and-computer-privacy/">Feistel, H. (1973)</a> Cryptography and Computer Privacy. <em>Scientific American</em>, 228(5).</li>
</ul>
</body>
</html>
