\documentclass{bioinfo}
\copyrightyear{2014}
\pubyear{2014}

\begin{document}
\firstpage{1}

\title{UniqTag: Assign unique, reasonably stable, content-derived
identifiers to genes}
\author{Shaun Jackman$^{1,2,*}$, Joerg Bohlmann$^2$ and Inanc Birol$^{1,2}$
\footnote{to whom correspondence should be addressed}}

\address{$^{1}$BC Cancer Agency Genome Sciences Centre
\\ $^{2}$University of British Columbia}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}

\section{Summary}

UniqTag assigns unique identifiers to gene sequences, or other arbitrary
sequences of characters, that are derived from the \emph{k}-mer
composition of the sequence. Unlike serial or accession numbers, these
identifiers are reasonably stable between different assemblies and
annotations of the same data.

\section{Availability and implementation}

The implementation of UniqTag is available at
\\ \texttt{https://github.com/sjackman/uniqtag}

Supplementary data and code to reproduce it is available at
\\ \texttt{https://github.com/sjackman/uniqtag-paper}

\section{Contact}

Shaun Jackman \textless{}sjackman@bcgsc.ca\textgreater{}

\end{abstract}

\section{Introduction}

The task of annotating the genes of a genome sequence often follows
genome sequence assembly. These annotated genes are assigned unique
identifiers by which they can be referenced. Assembly and annotation is
often an iterative process, by refining the method or by the addition of
the more sequencing data. These gene identifiers would ideally be
reasonably stable from one assembly to the next. Genes are typically
assigned serial or accession numbers, which, although certainly unique,
are not stable between assemblies. A single change in the assembly can
result in a total renumbering of the annotated genes.

One solution to stabilize identifiers is to assign them based on the
content of the gene sequence. A cryptographic hash function such as SHA
(Secure Hash Algorithm)
(\href{http://www.nist.gov/manuscript-publication-search.cfm?pub_id=910977}{Dang,
2012}) derives a message digest from the sequence, such that two
sequences with the same content will have the same message digest, and
two sequences that differ will have different message digests. If a
cryptographic hash were used to identify a gene, the same gene in two
assemblies with identical content would be assigned identical
identifiers, but by design a slight change in the sequence, such as a
single-character substitution, would result in a completely different
digest and identifier.

A cryptographic hash function is designed so that small changes in the
message, even a single bit change, results in large changes to the
message digest: half of the bits of the digest are expected to flip,
called the avalanche effect
(\href{http://www.scientificamerican.com/article/cryptography-and-computer-privacy/}{Feistel,
1973}). Locality-sensitive hashing (LSH) in contrast aims to assign
items that are similar to the same hash bucket. A hash function that,
after a small perturbation of the sequence, assigns an identical
identifier to the sequence is desirable for identifying the genes of a
genome sequence assembly project. One such locality-sensitive hash
function, MinHash, was employed in identifying web pages with similar
content (\href{http://dx.doi.org/10.1109/SEQUEN.1997.666900}{Broder,
1997}). UniqTag implements MinHash, where the set of elements of an item
is the \emph{k}-mer composition of the sequence, the hash function is
the identity function and the minimal element is the lexicographically
minimal sequence, to assign reasonably stable identifiers to genes.
These identifiers are intended for systematic identification, unique
within an assembly, rather than as a biological name, which is typically
assigned based on biological function or homology to orthologous genes.

\section{Algorithm}

The following symbols and terms are defined.

\begin{itemize}
\itemsep1pt\parskip0pt\parsep0pt
\item
  $\Sigma$ is an alphabet
\item
  \emph{s} and \emph{t} are strings, a sequence of symbols over $\Sigma$
\item
  \emph{S} is a set of strings over $\Sigma$
\item
  $\Sigma^k$ is the set of all strings over $\Sigma$ of length \emph{k}
\item
  $\min S$ is the lexicographically minimal string of \emph{S}
\item
  $C(s)$ is the set of all substrings of \emph{s}
\item
  A \emph{k}-mer of \emph{s} is a substring of \emph{s} with length
  \emph{k}, also called an \emph{n}-gram
\item
  $C_k(s)$ is the set of all \emph{k}-mers present in \emph{s}
\item
  $f(s, S)$ is the frequency of \emph{s} in \emph{S}, defined as the
  number of strings in \emph{S} that contain \emph{s} as a substring
\item
  $\mathop{\arg\,\min}\limits_{t \in C_k(s)} f(t, S)$ is the set of the
  \emph{k}-mers of \emph{s} that are least frequent in \emph{S}
\item
  $u_k(s, S)$ is the UniqTag, the lexicographically minimal \emph{k}-mer
  of the \emph{k}-mers of \emph{s} that are least frequent in \emph{S}
\end{itemize}

\clearpage
The UniqTag $u_k(s, S)$ is defined as follows.

\[
\begin{aligned}
C_k(s) &= C(s) \cap \Sigma^k
\\ f(s, S) &= \left\vert \{ t \mid s \in C(t) \wedge t \in S \} \right\vert
\\ u_k(s, S) &= \min \mathop{\arg\,\min}\limits_{t \in C_k(s)} f(t, S)
\end{aligned}
\]

\section{Results}

UniqTag was used to assign identifiers to the protein sequences of six
builds of the Ensembl human genome
(\href{http://dx.doi.org/10.1093/nar/gkt1196}{Flicek, 2014}) spanning
five years. The overlap of UniqTag identifers between older builds and
the current build 75 is shown in Table 1.

\begin{table}[!h]
\processtable{The overlap of UniqTag identifers between older builds
of the Ensembl human genome and the current build 75.}
{
\begin{tabular}{rrrrr}
\hline\noalign{\medskip}
Build A & Build B & Only in A & In both & Only in B
\\\noalign{\medskip}
\hline\noalign{\medskip}
55 & 75 & 24299 & 30997 & 17600
\\\noalign{\medskip}
60 & 75 & 9365 & 34859 & 13738
\\\noalign{\medskip}
65 & 75 & 1088 & 45931 & 2666
\\\noalign{\medskip}
70 & 75 & 231 & 47955 & 642
\\\noalign{\medskip}
74 & 75 & 0 & 48597 & 0
\\\noalign{\medskip}
\hline
\noalign{\medskip}
\end{tabular}
}{}
\end{table}

\section{Discussion}

When iterating over multiple assemblies of the same data, it is rather
inconvenient when gene identifiers to change from one assembly to the
next. UniqTag attempts to address this common annoyance. By identifying
the gene by a feature of its content rather than an arbitrary serial
number, the gene identifier is reasonably stable between assemblies.

A UniqTag will change due to a difference in the locus of the UniqTag
itself, the creation of a least-frequent \emph{k}-mer that is
lexicographically smaller than the previous UniqTag, or the creation of
a \emph{k}-mer elsewhere resulting in the UniqTag no longer being a
least-frequent \emph{k}-mer. Concatenating two gene models results in a
gene whose UniqTag is the minimum of the two previous UniqTags, unless
one of the k-mer at the junction of the two sequences is
lexicographically smaller. Similarly, splitting a gene model in two
results in one gene that retains the previous UniqTag and a second gene
that is assigned a new UniqTag, unless the previous UniqTag spanned the
junction.

A UniqTag can be generated from the nucleotide sequence of a gene. Using
instead the translated amino acid sequence of a protein-coding gene
sequence results in a UniqTag that is stable across synonymous changes
to the coding sequence as well as to changes in the untranslated regions
(UTR) and introns of the gene. Since the amino acid alphabet is larger
than the nucleotide alphabet, fewer characters are required for a
\emph{k}-mer to be likely unique, resulting in an aesthetically pleasing
shorter identifier.

Two gene models with identical sequence will have the same UniqTag. It
is possible although uncommon that two non-identical genes with similar
\emph{k}-mer composition and no unique \emph{k}-mer are assigned the
same UniqTag. Genes that are assigned the same UniqTag are distinguished
by adding a numerical suffix to the UniqTag.

\section*{Acknowledgement}

Thanks to Nathaniel Street for his enthusiastic feedback, to the
SMarTForests project, its funders and the organizers of the 2014 Conifer
Genome Summit that made our conversation possible.

\paragraph{Funding\textcolon}
This work was supported by the Natural Sciences and Engineering Research
Council of Canada, Genome British Columbia, Genome Alberta, Genome
Qu√©bec and Genome Canada.

\section*{References}

\begin{itemize}
\itemsep1pt\parskip0pt\parsep0pt
\item[]
  \href{http://dx.doi.org/10.1109/SEQUEN.1997.666900}{Broder, A. Z.
  (1997)} On the resemblance and containment of documents.
  \emph{Compression and Complexity of Sequences}, 1997 Proceedings,
  21-29.
\item[]
  \href{http://www.nist.gov/manuscript-publication-search.cfm?pub_id=910977}{Dang,
  Q. H. (2012)} Secure Hash Standard (SHS). \emph{NIST FIPS}, 180(4),
  1-35.
\item[]
  \href{http://www.scientificamerican.com/article/cryptography-and-computer-privacy/}{Feistel,
  H. (1973)} Cryptography and Computer Privacy. \emph{Scientific
  American}, 228(5).
\item[]
  \href{http://dx.doi.org/10.1093/nar/gkt1196}{Flicek, P. (2014)}
  Ensembl 2014. \emph{Nucleic Acids Research}, 42(D1), D749-D755.
\end{itemize}
\end{document}
